src/internal/SlidingSheet.js

```diff
@@ -0,0 +1,439 @@
+    const direction = this.props.direction;
+    return direction === 'right' || direction === 'bottom' ? 1 : -1;
+  }
+
+  enableSwipeHandling() {
+    if (this.props.swipeToOpen || this.props.closeable) {
+      document.body.addEventListener('touchstart', this.onBodyTouchStart);
+      if (this.props.swipeToOpen && !openNavEventHandler) {
+        openNavEventHandler = this.onBodyTouchStart;
+      }
+    } else {
+      this.disableSwipeHandling();
+    }
+  }
+
+  disableSwipeHandling() {
+    document.body.removeEventListener('touchstart', this.onBodyTouchStart);
+    if (openNavEventHandler === this.onBodyTouchStart) {
+      openNavEventHandler = null;
+    }
+  }
+
+  onBodyTouchStart = (event) => {
+    const swipeAreaWidth = this.props.swipeAreaWidth;
+    const direction = this.props.direction;
+
+    const touchStartX = event.touches[0].pageX;
+    const touchStartY = event.touches[0].pageY;
+
+    const vertical = this.isVertical();
+
+    const touchStartPrimary = vertical ? touchStartY : touchStartX;
+    const touchStartSecondary = vertical ? touchStartX : touchStartY;
+
+    const bodyPrimaryDimension = vertical ? document.body.offsetHeight : document.body.offsetWidth;
+
+    // Open only if swiping from the edge while closed
+    if (swipeAreaWidth !== null && !this.props.open) {
+      if (direction === 'right' || direction === 'bottom') {
+        // Calculate from the bottom/right side
+        if (touchStartPrimary < bodyPrimaryDimension - swipeAreaWidth) return;
+      } else {
+        // Calculate from the top/left side
+        if (touchStartPrimary > swipeAreaWidth) return;
+      }
+    }
+
+    if (!this.props.open &&
+         (openNavEventHandler !== this.onBodyTouchStart ||
+          !this.props.swipeToOpen) ||
+         this.props.open && !this.props.closeable) {
+      return;
+    }
+
+    // For a vertical sheet, only start swiping away if the user
+    // is touching the edge of the sheet
+    if (this.props.open && vertical) {
+      const sheetHeight = ReactDOM.findDOMNode(this.refs.clickAwayableElement).offsetHeight;
+      const sheetEdge = direction === 'top' ? sheetHeight : window.innerHeight - sheetHeight;
+      const windowTouchStartY = touchStartY - window.pageYOffset;
+
+      if (windowTouchStartY >= sheetEdge - 30 && windowTouchStartY <= sheetEdge + 30) {
+        // We can start swiping down right away, rather than setting this.maybeSwiping
+        this.swipeStartPrimary = touchStartPrimary;
+        this.setState({swiping: 'closing'});
+      } else {
+        return;
+      }
+    } else {
+      this.maybeSwiping = true;
+    }
+
+    this.touchStartPrimary = touchStartPrimary;
+    this.touchStartSecondary = touchStartSecondary;
+
+    document.body.addEventListener('touchmove', this.onBodyTouchMove);
+    document.body.addEventListener('touchend', this.onBodyTouchEnd);
+    document.body.addEventListener('touchcancel', this.onBodyTouchEnd);
+  };
+
+  setPosition(translatePrimary) {
```

Comment:
*Side comment that is out of the scope of this PR*:
Notice that `setPosition` can potentially be triggered at a higher frequency than the screen refresh rate (60Hz).
See https://github.com/callemall/material-ui/issues/2862 as reference.